### PR TITLE
basic font install support

### LINKED ISSUE #16: Add a way to upload a TTF file and install the font system wide

### PR DESCRIPTION
**Pull request by nonprofittechy**
Fix #16

Feel free to skip review if you don't have time.

### CODE REVIEW

File: docassemble/ALDashboard/aldashboard.py [hunk 1]
```diff
@@ -241,6 +246,35 @@ def init(self, *pargs, **kwargs):
...
+    if get_config("enable unoconv"):
+        output += subprocess.run(["supervisorctl", "-s", "http://localhost:9001", "restart", "unoconv"], capture_output=True, text=True).stdout
+
+    return output
+
+def list_installed_fonts():
+    """
+    List the fonts installed on the server.
+    """
+    output = subprocess.run(["fc-list"], capture_output=True, text=True).stdout
```
Comment (BryceStevenWilley) [inline docassemble/ALDashboard/aldashboard.py:274]: The default listing isn't in any sort of order (which is weird IMO). I'd suggest piping the output through `sort`. 
```suggestion
    fc_list = subprocess.run(["fc-list"], stdout=subprocess.PIPE)
    output = subprocess.run(["sort"], stdin=fc_list.stdout, capture_output=True, text=True).stdout
    fc_list.stdout.close()
```

From https://docs.python.org/3/library/subprocess.html#replacing-shell-pipeline

File: docassemble/ALDashboard/data/questions/install_fonts.yml [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+  ${ action_button_html(url_action("list_fonts_task"), label="View installed fonts", id_tag="view_fonts") }
+
+  [TARGET displayed_fonts]
+
+fields:
+  - Font files: the_font_files
+    datatype: files
+    #accept: |
+    #  "application/font-sfnt, .TTF, .ttf, font/*"
```
Comment (BryceStevenWilley) [inline docassemble/ALDashboard/data/questions/install_fonts.yml:70]: Was this not working for some reason? Was going to suggest it, only seeing the files you're allowed to upload really helps.

File: docassemble/ALDashboard/data/questions/install_fonts.yml [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  list_fonts_task = background_action("display_fonts", 'backgroundresponse')
+---
+event: display_fonts
+code: |
+  installed_fonts = f"""
+    <div class="alert alert-info">
+      <h4>Installed fonts</h4>
+        { list_installed_fonts() }
+    </div>
+  """
```
Comment (BryceStevenWilley) [inline docassemble/ALDashboard/data/questions/install_fonts.yml:86]: It's a little confusing to click the "List installed fonts" button and then have nothing happen for about 10 seconds. It'd be nice if `displayed_fonts` popped up empty at first or something, but I can't see an easy way to do that with DA background stuff. :/

File: docassemble/ALDashboard/aldashboard.py [hunk 4]
```diff
@@ -241,6 +246,35 @@ def init(self, *pargs, **kwargs):
...
+    """
+    Install fonts to the server and restart both supervisor and unoconv.
+    """
+    # create the /var/www/.fonts directory if it doesn't exist
+    if not os.path.exists("/var/www/.fonts"):
+        os.makedirs("/var/www/.fonts")
+    
+    # save the DAFile to /var/www/.fonts
+    for f in the_font_files:
+        shutil.copyfile(f.path(), "/var/www/.fonts/" + f.filename)
```
Comment (BryceStevenWilley) [inline docassemble/ALDashboard/aldashboard.py:259]: I'd be fairly worried about some sort of injection attack here (even though it's an admin only interview, idk). DA already uses `werkzeug.secure_filename()`, that should be a drop-in fix. (Might also have to import it in addition to this edit). 

```suggestion
        shutil.copyfile(f.path(), "/var/www/.fonts/" + werkzeug.secure_filename(f.filename))
```

**Review by BryceStevenWilley — APPROVED** [review]
LGTM. 

Not sure if you've seen https://demo.docassemble.org/run/demo/fontlist/#/1, but TBH, it's a bit more readable than the component you have here, even though it's still useful (have to think of and rely on in-page search).

I also can't seem to actually use my font once it's installed (trying to follow https://docassemble.org/docs/documents.html#rendering%20font). But that's either a me problem, or a docassemble problem, so it shouldn't block this.

File: docassemble/ALDashboard/data/questions/install_fonts.yml (hunk 2 above)
Comment (nonprofittechy) [reply docassemble/ALDashboard/data/questions/install_fonts.yml:70]: It didn't work. When it was enabled, it was blocking any upload of a TTF straight from c:\windows\fonts

File: docassemble/ALDashboard/data/questions/install_fonts.yml (hunk 3 above)
Comment (nonprofittechy) [reply docassemble/ALDashboard/data/questions/install_fonts.yml:86]: Agreed. I tried hooking a JS event there to change the button status, but after trying the obvious way and it failing: I am OK with this being a little rough around the edges for now.

**Comment by nonprofittechy** [comment]
> LGTM.
> 
> Not sure if you've seen https://demo.docassemble.org/run/demo/fontlist/#/1, but TBH, it's a bit more readable than the component you have here, even though it's still useful (have to think of and rely on in-page search).

Hmm. That is way better actually. I don't know that I'll change it now though, but I'll open an issue to switch it at some point. 

> I also can't seem to actually use my font once it's installed (trying to follow https://docassemble.org/docs/documents.html#rendering%20font). But that's either a me problem, or a docassemble problem, so it shouldn't block this.

It definitely works when I upload a new PDF with the font. It may only work in Libreoffice, because of the directory we install the font to.

**Pull request merged**